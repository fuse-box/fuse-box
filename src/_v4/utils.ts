import * as fs from "fs";
import * as fsExtra from "fs-extra";
import { ensurePublicExtension, extractFuseBoxPath } from "../Utils";

const CACHED_PATHS = new Map<string, RegExp>();
export function path2Regex(path: string) {
	if (CACHED_PATHS.get(path)) {
		return CACHED_PATHS.get(path);
	}
	path = path.replace(/(\.|\/)/, "\\$1");

	const re = new RegExp(path);
	CACHED_PATHS.set(path, re);
	return re;
}

export function fileExists(file: string) {
	return fs.existsSync(file);
}

export function ensureDir(dir: string) {
	fsExtra.ensureDirSync(dir);
}
export function makeFuseBoxPath(homeDir: string, absPath: string) {
	return homeDir && ensurePublicExtension(extractFuseBoxPath(homeDir, absPath));
}
