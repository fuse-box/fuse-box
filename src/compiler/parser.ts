import * as parser from '@typescript-eslint/typescript-estree';
import * as meriyah from 'meriyah';
import { ASTNode } from './interfaces/AST';
export interface IParserOptions {
  jsx?: boolean;
  locations?: boolean;
}

export type ICodeParser = (code: string, props?: IParserOptions) => ASTNode;

export function parseTypeScript(code: string, props?: IParserOptions): ASTNode {
  props = props || {};
  return parser.parse(code, {
    range: props.locations,
    raw: true,
    useJSXTextNode: true,
    loc: props.locations,
    jsx: props.jsx,
  }) as any;
}

export function parseJavascript(code: string, props?: IParserOptions): ASTNode {
  props = props || {};

  return meriyah.parse(code, { jsx: props.jsx, loc: props.locations, module: true }) as any;
}
