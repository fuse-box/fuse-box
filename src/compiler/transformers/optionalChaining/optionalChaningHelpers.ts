import { ASTNode, ASTType } from '../../interfaces/AST';

export function ensureExpressionConsistency(expression: ASTNode, node?: ASTNode) {
  if (node.type == ASTType.Literal) {
    expression.computed = true;
  }
}

export function convertToCallExpression(alternate: ASTNode, id, args): ASTNode {
  return {
    arguments: args,
    callee: { object: { name: id, type: 'Identifier' }, property: alternate, type: 'MemberExpression' },
    type: 'CallExpression',
  };
}

export type OptionalChainHelper = ReturnType<typeof createOptionalChaningExpression>;
export function createOptionalChaningExpression(userId: string) {
  // prevMember.test.left.left.right

  // prevMement.alternate.property
  const AST_NODE: ASTNode = {
    alternate: null,
    consequent: {
      argument: {
        type: 'Literal',
        value: 0,
      },
      operator: 'void',
      prefix: true,
      type: 'UnaryExpression',
    },
    test: {
      left: {
        left: {
          left: {
            name: userId,
            type: 'Identifier',
          },
          operator: '=',
          right: null,
          type: 'AssignmentExpression',
        },
        operator: '===',
        right: {
          type: 'Literal',
          value: null,
        },
        type: 'BinaryExpression',
      },
      operator: '||',
      right: {
        left: {
          name: userId,
          type: 'Identifier',
        },
        operator: '===',
        right: {
          argument: {
            type: 'Literal',
            value: 0,
          },
          operator: 'void',
          prefix: true,
          type: 'UnaryExpression',
        },
        type: 'BinaryExpression',
      },
      type: 'LogicalExpression',
    },
    type: 'ConditionalExpression',
  };

  return {
    id: userId,
    statement: AST_NODE,
    setLeft: (item: ASTNode) => {
      AST_NODE.test.left.left.right = item;
    },
    setRight: (item: ASTNode) => {
      AST_NODE.alternate = item;
    },
  };
}
