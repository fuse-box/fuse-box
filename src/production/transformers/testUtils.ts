import { ITransformer } from '../../compiler/interfaces/ITransformer';
import * as buntis from 'buntis';
import { GlobalContextTransformer } from '../../compiler/transformers/GlobalContextTransformer';
import { transpileModule } from '../../compiler/program/transpileModule';
import { createGlobalContext } from '../../compiler/program/GlobalContext';
import { createContext } from '../../core/Context';
import { createPackage } from '../../core/Package';
import { createModule } from '../../core/Module';
import { ASTNode } from '../../compiler/interfaces/AST';
import { ProductionContext } from '../ProductionContext';
import { ModuleTree } from '../module/ModuleTree';

export function testProductionWarmup(props: {
  jsx?: boolean;
  code: string;
  transformers: Array<ITransformer>;
  props?: any;
  moduleProps?: any;
}) {
  const ctx = createContext({
    cache: false,
    devServer: false,
    target: 'browser',
  });

  const pkg = createPackage({ ctx: ctx, meta: {} as any });

  const module = Object.assign(
    createModule({
      absPath: __filename,
      ctx: ctx,
      extension: '.ts',
      fuseBoxPath: 'file.ts',
    }, pkg),
    props.moduleProps || {},
  );

  pkg.modules.push(module);

  module.contents = props.code;
  module.parse();

  const productionContext = ProductionContext(ctx, [pkg]);

  const tree = (module.moduleTree = ModuleTree(productionContext, module));

  const tranformers = [GlobalContextTransformer().commonVisitors(props.props)];
  for (const t of props.transformers) {
    if (t.productionWarmupPhase) {
      tranformers.push(t.productionWarmupPhase({ ctx: ctx, module: module, productionContext: productionContext }));
    }
  }
  transpileModule({
    ast: module.ast as ASTNode,
    globalContext: createGlobalContext(),
    transformers: tranformers,
  });

  return { productionContext, module, tree };
}
