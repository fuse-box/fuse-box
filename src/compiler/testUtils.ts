import { ICompilerOptions } from '../compilerOptions/interfaces';
import { transformModule } from './core/transformModule';
import { generate } from './generator/generator';
import { ASTNode } from './interfaces/AST';
import { ITransformer, ITransformerCommon } from './interfaces/ITransformer';
import { ITransformerRequireStatementCollection } from './interfaces/ITransformerRequireStatements';
import { ImportType } from './interfaces/ImportType';
import { parseTypeScript } from './parser';
import { ISerializableTransformationContext } from './transformer';

function cleanupForTest(node) {
  delete node.loc;
  delete node.raw;
  delete node.range;
}
export function initCommonTransform(props: {
  code: string;
  compilerOptions?: ICompilerOptions;
  jsx?: boolean;
  props?: ISerializableTransformationContext;
  transformers: Array<ITransformer>;
}) {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      cleanupForTest(statement.arguments[0]);
      cleanupForTest(statement);
      cleanupForTest(statement.callee);

      requireStatementCollection.push({ importType, statement });
    }
  }

  const ast = parseTypeScript(props.code, { jsx: props.jsx });
  const userProps: ISerializableTransformationContext = props.props || {};
  userProps.compilerOptions = props.compilerOptions || {};
  const visitorProps: ITransformerCommon = { onRequireCallExpression, transformationContext: userProps };

  const tranformers = [];
  for (const t of props.transformers) {
    if (t.commonVisitors) {
      tranformers.push(t.commonVisitors(visitorProps));
    }
  }

  transformModule({ root: ast, transformers: tranformers });

  const res = generate(ast, {});

  return { code: res, requireStatementCollection };
}
