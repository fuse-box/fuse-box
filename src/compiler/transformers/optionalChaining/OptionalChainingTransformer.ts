import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { createUndefinedVariable, createVariableDeclaration } from '../../helpers/astHelpers';
import { ASTNode, ASTType } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { OptionalChainHelper, createOptionalChaningExpression } from './optionalChaningHelpers';

const VALID_NODES = {
  [ASTType.OptionalCallExpression]: 1,
  [ASTType.OptionalMemberExpression]: 1,
};

const NON_DRILLABLE_NODES = {
  [ASTType.CallExpression]: 1,
  [ASTType.Identifier]: 1,
  [ASTType.MemberExpression]: 1,
};

interface IChainingStep {
  callArguments?: Array<ASTNode>;
  computed?: boolean;
  expression?: ASTNode;
  optional?: boolean;
}

interface IChainItem {
  items: Array<IChainingStep>;
}

type IChainingCollection = Array<IChainItem>;

function createOptionalContext(visit: IVisit) {
  const steps: Array<IChainingStep> = [];
  const globalContext = visit.globalContext;
  let finalStatement: ASTNode;
  let flatCollection: IChainingCollection;
  const declaration = createVariableDeclaration();

  const self = {
    declaration,
    finalStatement,
    flatCollection,
    steps,
    genId: () => {
      const nextVar = globalContext.getNextSystemVariable();
      declaration.declarations.push(createUndefinedVariable(nextVar));
      return nextVar;
    },
  };
  return self;
}
type OptionalChainContext = ReturnType<typeof createOptionalContext>;

function createPartialExpression(props: {
  flatExpression: IFlatExpression;
  nodes: Array<IChainingStep>;
  shouldExtractThis: boolean;
}) {
  let expression: ASTNode;
  const { flatExpression, nodes, shouldExtractThis } = props;
  const firstNode = nodes[0];
  let thisVariable: string;

  if (nodes.length === 1) {
    if (shouldExtractThis) {
      if (firstNode.expression.type === 'MemberExpression') {
        thisVariable = flatExpression.context.genId();
        firstNode.expression.object = {
          left: { name: thisVariable, type: 'Identifier' },
          operator: '=',
          right: firstNode.expression.object,
          type: 'AssignmentExpression',
        };
        return { expression: firstNode.expression, thisVariable };
      }
    }
    return { expression: firstNode.expression };
  }

  const total = nodes.length;
  let index = 0;

  while (index < total) {
    const item = nodes[index];
    const isLast = index === total - 1;

    const computed = item.computed || (item.expression && item.expression.type === ASTType.Literal);

    // call expression
    if (item.callArguments) {
      if (index === 1) {
        expression = createOptionalCall(flatExpression, item.callArguments);
      } else {
        expression = {
          arguments: item.callArguments,
          callee: expression,
          type: 'CallExpression',
        };
      }
    } else {
      if (!expression) {
        expression = {
          object: item.expression,
          property: null,
          type: 'MemberExpression',
        };
      } else {
        // if we see a next property while having an existing CallExpression, we must
        // converte to a MemberExpression with an object as the current CallExpression
        if (expression.type === 'CallExpression')
          expression = { object: expression, property: null, type: 'MemberExpression' };

        if (!expression.property) {
          expression.property = item.expression;
          expression.computed = computed;
        } else {
          expression = {
            computed: computed,
            object: expression,
            property: item.expression,
            type: 'MemberExpression',
          };
        }
      }
    }
    if (isLast && shouldExtractThis) {
      // adding an expression statement
      // E.g a statement like a?.b.c.d.e.f?.();
      // The second part b.c.d.e.f, gets converted into
      // (_new_variable = _1_.b.c.d.e).f)
      if (expression.object) {
        thisVariable = flatExpression.context.genId();
        expression.object = {
          left: { name: thisVariable, type: 'Identifier' },
          operator: '=',
          right: expression.object,
          type: 'AssignmentExpression',
        };
      } else {
      }
    }
    index++;
  }

  return { expression, thisVariable };
}

interface IFlatExpression {
  alternate?: Array<IChainingStep>;
  conditionSteps?: Array<IChainingStep>;
  context?: OptionalChainContext;
  id?: string;
  nextFlatExpression?: IFlatExpression;
  shouldExtractThis?: boolean;
  thisVariable?: string;
  collect?: () => OptionalChainHelper;
  generate?: () => OptionalChainHelper;
}

function createOptionalCall(expression: IFlatExpression, callArguments: Array<ASTNode>): ASTNode {
  let args: Array<ASTNode> = [];
  if (expression.thisVariable) {
    args.push({
      name: expression.thisVariable,
      type: 'Identifier',
    });
  } else {
    return {
      arguments: callArguments,
      callee: {
        computed: false,
        name: expression.id,
        type: 'Identifier',
      },
      type: 'CallExpression',
    };
  }
  args = args.concat(callArguments);
  return {
    arguments: args,
    callee: {
      computed: false,
      object: {
        name: expression.id,
        type: 'Identifier',
      },
      property: {
        name: 'call',
        type: 'Identifier',
      },
      type: 'MemberExpression',
    },
    type: 'CallExpression',
  };
}

function createFlatExpression(context: OptionalChainContext): IFlatExpression {
  const self: IFlatExpression = {
    context,
    collect: (): OptionalChainHelper => {
      let initialLeft;

      if (self.conditionSteps) {
        initialLeft = createPartialExpression({
          flatExpression: self,
          nodes: self.conditionSteps,
          shouldExtractThis: !!self.alternate[0].callArguments,
        });
      }

      self.id = context.genId();
      const targets = self.alternate;
      let expression = createOptionalChaningExpression(self.id);

      if (self.conditionSteps) {
        expression.setLeft(initialLeft.expression);
        if (initialLeft.thisVariable) {
          self.thisVariable = initialLeft.thisVariable;
        }
      }

      // inserting the current identifier

      targets.unshift({ expression: { name: self.id, type: 'Identifier' } });
      const dataRight = createPartialExpression({
        flatExpression: self,
        nodes: targets,
        shouldExtractThis: self.shouldExtractThis,
      });
      expression.setRight(dataRight.expression);

      // should notify the following call expression to use "this" variable
      if (self.nextFlatExpression && dataRight.thisVariable)
        self.nextFlatExpression.thisVariable = dataRight.thisVariable;

      return expression;
    },

    generate: (): OptionalChainHelper => {
      let expression = self.collect();
      let next = self.nextFlatExpression;
      while (next) {
        const data = next.collect();
        data.setLeft(expression.statement);
        expression = data;
        next = next.nextFlatExpression;
      }
      return expression;
    },
  };
  return self;
}

function processFlatCollection(context: OptionalChainContext) {
  const flatCollection = context.flatCollection;
  let current: IFlatExpression;

  let index = flatCollection.length - 1;
  while (index >= 0) {
    const item = flatCollection[index];
    if (index > 0) {
      let flatExpression: IFlatExpression = createFlatExpression(context);
      flatExpression.alternate = item.items;
      if (current) {
        flatExpression.shouldExtractThis = !!current.alternate[0].callArguments;
        flatExpression.nextFlatExpression = current;
      }
      current = flatExpression;
    } else {
      current.conditionSteps = item.items;
    }
    index--;
  }

  const { statement } = current.generate();
  context.finalStatement = statement;
}

function flatten(context: OptionalChainContext) {
  const { steps } = context;
  const amount = steps.length;

  const collection: IChainingCollection = [];

  let index = amount - 1;
  while (index >= 0) {
    const step = steps[index];
    const prev = collection[collection.length - 1];
    if (!step.optional && prev) prev.items.push(step);
    else collection.push({ items: [step] });
    index--;
  }
  context.flatCollection = collection;
  processFlatCollection(context);
}
/**
 * Drill every single property on the OptionalChain
 * Split it into steps and prepare for flattening
 * @param node
 * @param context
 */
export function chainDrill(node: ASTNode, context: OptionalChainContext) {
  const optional = node.optional === true;

  if (NON_DRILLABLE_NODES[node.type]) {
    return context.steps.push({ computed: node.computed, expression: node });
  }

  if (node.type === ASTType.OptionalMemberExpression) {
    if (node.property) context.steps.push({ computed: node.computed, expression: node.property, optional });
    if (node.object) {
      if (node.object.type === ASTType.CallExpression) {
        context.steps.push({ expression: node.object, optional });
      } else chainDrill(node.object, context);
    }
    return;
  }

  // Regardless of the "optional" status calls should be separated since the call to those
  if (node.type === ASTType.OptionalCallExpression) {
    if (node.callee) {
      context.steps.push({
        callArguments: node.arguments,
        optional,
      });
      chainDrill(node.callee, context);
    }
  }
}

export function OptionalChaningTransformer(): ITransformer {
  return {
    commonVisitors: props => {
      return {
        onEachNode: (visit: IVisit): IVisitorMod => {
          const node = visit.node;
          if (!VALID_NODES[node.type]) return;
          const context = createOptionalContext(visit);

          chainDrill(node, context);
          flatten(context);

          return { prependToBody: [context.declaration], replaceWith: context.finalStatement };
        },
      };
    },
  };
}
