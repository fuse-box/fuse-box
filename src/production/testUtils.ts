import * as appRoot from 'app-root-path';
import { unlinkSync, writeFileSync, unlink } from 'fs';
import * as path from 'path';
import { ASTNode } from '../compiler/interfaces/AST';
import { ITransformer } from '../compiler/interfaces/ITransformer';
import { createGlobalContext } from '../compiler/program/GlobalContext';
import { transpileModule } from '../compiler/program/transpileModule';
import { GlobalContextTransformer } from '../compiler/transformers/GlobalContextTransformer';
import { IPublicConfig } from '../config/IPublicConfig';
import { createContext } from '../core/Context';
import { fusebox } from '../core/fusebox';
import { createModule } from '../core/Module';
import { createPackage } from '../core/Package';
import { ensureDir, fastHash } from '../utils/utils';
import { ModuleTree } from './module/ModuleTree';
import { IProductionContext, ProductionContext } from './ProductionContext';
import * as fsExtra from 'fs-extra';
export function testProductionWarmup(props: {
  jsx?: boolean;
  code: string;
  transformers: Array<ITransformer>;
  props?: any;
  moduleProps?: any;
}) {
  const ctx = createContext({
    cache: false,
    devServer: false,
    target: 'browser',
  });

  const pkg = createPackage({ ctx: ctx, meta: {} as any });

  const module = Object.assign(
    createModule(
      {
        absPath: __filename,
        ctx: ctx,
        extension: '.ts',
        fuseBoxPath: 'file.ts',
      },
      pkg,
    ),
    props.moduleProps || {},
  );

  pkg.modules.push(module);

  module.contents = props.code;
  module.parse();

  const productionContext = ProductionContext(ctx, [pkg]);

  const tree = (module.moduleTree = ModuleTree(productionContext, module));

  const tranformers = [GlobalContextTransformer().commonVisitors(props.props)];
  for (const t of props.transformers) {
    if (t.productionWarmupPhase) {
      tranformers.push(t.productionWarmupPhase({ ctx: ctx, module: module, productionContext: productionContext }));
    }
  }
  transpileModule({
    ast: module.ast as ASTNode,
    globalContext: createGlobalContext(),
    transformers: tranformers,
  });

  return { productionContext, module, tree };
}

function createFileSet(directory, files: Record<string, string>) {
  const newSet = {};
  for (const key in files) {
    newSet[path.join(directory, key)] = files[key];
  }

  return newSet;
}
export class ProdPhasesTestEnv {
  private rootDir: string;
  private homeDir: string;
  constructor(public config: IPublicConfig) {
    config.cache = false;
    config.target = 'browser';
    config.watch = false;
    config.devServer = false;
  }
  initFileStructure(files: Record<string, string>) {
    const folder = fastHash(Math.random() + '-' + new Date().getTime());
    this.rootDir = path.join(appRoot.path, '.tmp', folder);
    this.homeDir = path.join(this.rootDir, 'src');
    this.config.homeDir = this.homeDir;
    ensureDir(this.homeDir);
    for (const key in files) {
      const filePath = path.join(this.homeDir, key);
      writeFileSync(filePath, files[key]);
    }
  }

  public async run(phases): Promise<IProductionContext> {
    const fuse = fusebox(this.config);
    const context = await fuse.runProductionContext(phases);
    return context;
  }

  public destroy() {
    fsExtra.removeSync(this.rootDir);
  }
}

export function prodPhasesEnv(config: IPublicConfig, files: Record<string, string>): ProdPhasesTestEnv {
  const env = new ProdPhasesTestEnv(config);
  env.initFileStructure(files);

  return env;
}
