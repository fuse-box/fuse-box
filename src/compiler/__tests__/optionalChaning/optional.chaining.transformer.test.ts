import { writeFileSync } from 'fs';
import * as path from 'path';
import * as ts from 'typescript';
import { fileExists, readJSONFile } from '../../../utils/utils';
import { generate } from '../../generator/generator';
import { parseJavascript } from '../../parser';
import { initCommonTransform } from '../../testUtils';
import { OptionalChaningTransformer } from '../../transformers/optionalChaining/OptionalChainingTransformer';

const testTranspile = (props: { code: string; jsx?: boolean }) => {
  return initCommonTransform({
    code: props.code,
    jsx: true,
    props: { module: {} },
    transformers: [OptionalChaningTransformer()],
  });
};

const cases: Array<string> = [
  'a?.b',
  'a?.b?.c',
  'a?.b.c',
  'a?.b?.c?.d',
  'a?.b?.c?.d?.e',
  'a?.b?.c?.d?.e?.f',
  'a.b?.c?.d?.e?.f',
  'a.b.c?.d?.e?.f',
  'a.b.c.d?.e?.f',
  'a.b?.c.d?.e.f',
  'a.b?.c.d?.e?.f',

  // handle computed properties
  'a?.[0]',
  'a?.c[0]',
  'a?.c?.[0]',
  'a?.["foo"]',
  'a?.c["foo"]',
  'a?.c?.["foo"]?.d',
  'a.c?.["foo"]?.d.e?.f',

  // optional calls set 1
  'a?.()',
  'a?.b?.()',
  'a?.b.c?.()',
  'a?.b.c?.(1,2,3,4)',
  'a?.b.c?.d.e?.()',
  'a?.b.c?.d.e?.().g()',

  // optional calls set 2
  'a?.()?.()',
  'a?.()?.a.b?.c',
  'a?.().a.k?.c',
  'a?.()?.(1)(2)(3)',
  'a?.()?.(1)[foo.bar]',
  'a?.()?.(1)?.(2)?.(3)',
  'a?.()?.(1)?.(2)?.(3)?.a.b?.()()',
  'a?.()?.(1)?.(2)?.(3)?.a["b"]?.()()',

  // optional calls set 3
  'a.b.c().d?.e?.()',
  'a.b.c().d?.e?.().f()?.()',
  'a.b.c()?.[d.e]?.["one"]()?.f()?.()',

  // starting expressions
  'a.b()?.().a',
  'a["foo"]?.();',
  'a["foo"]?.()?.()()',
  'a.b()?.()?.()',
  'a["foo"]()?.o()?.().k()?.c',
];

const TYPESCRIPT_SNAPSHOT_FILE = path.join(__dirname, 'typescript.snapshot.json');
let TYPESCRIPT_SHAPSHOTS;
function createTypeScriptSnapshots(cases: Array<string>) {
  TYPESCRIPT_SHAPSHOTS = {};
  for (const str of cases) {
    const compilerOptions: ts.CompilerOptions = {
      module: ts.ModuleKind.CommonJS,
      target: ts.ScriptTarget.ES2019,
    };
    const code = ts.transpileModule(str, {
      compilerOptions,
      fileName: __filename,
    });
    let transpiled = code.outputText;
    // replace typescript _a with fusebox numbers
    let alphabet = ['-', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k'];

    transpiled = transpiled.replace(/_([a-z])/gm, input => {
      const varName = input.split('_')[1];
      return `_${alphabet.indexOf(varName)}_`;
    });

    const ast = parseJavascript(transpiled);
    TYPESCRIPT_SHAPSHOTS[str] = generate(ast, {});
  }
  writeFileSync(TYPESCRIPT_SNAPSHOT_FILE, JSON.stringify(TYPESCRIPT_SHAPSHOTS, null, 2));
}

if (!fileExists(TYPESCRIPT_SNAPSHOT_FILE)) {
  createTypeScriptSnapshots(cases);
} else {
  TYPESCRIPT_SHAPSHOTS = readJSONFile(TYPESCRIPT_SNAPSHOT_FILE);
}

describe('Optional chaining', () => {
  describe('Snapshot test', () => {
    for (const str of cases) {
      it(`Should  handle ${str}`, () => {
        const res = testTranspile({ code: str });
        expect(res.code).toEqual(TYPESCRIPT_SHAPSHOTS[str]);
      });
    }
  });

  describe('Corner cases', () => {
    it('should handle asExpression', () => {
      const res = testTranspile({
        code: `
        (a as string)?.b?.c;
        a?.b?.c;
        `,
      });
      expect(res).toMatchSnapshot();
    });

    it('should await expression 1', () => {
      const res = testTranspile({
        code: `
        async function main(){
          await a?.b
        }
        `,
      });

      expect(res.code).toMatchSnapshot();
    });

    it('should await expression 2', () => {
      const res = testTranspile({
        code: `
        async function main(){
          await a?.b()
        }
        `,
      });

      expect(res.code).toMatchSnapshot();
    });

    it('should await expression with ?? ', () => {
      const res = testTranspile({
        code: `
        let b = async (a, c) => {
          const b = await a?.a() ?? await c?.d;
        };
        `,
      });

      expect(res.code).toMatchSnapshot();
    });
  });
});
