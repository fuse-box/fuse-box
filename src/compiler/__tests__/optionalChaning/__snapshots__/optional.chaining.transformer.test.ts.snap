// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Optional chaining Corner cases should await expression 1 1`] = `
"var _1_;
async function main() {
  await((_1_ = a) === null || _1_ === void 0 ? void 0 : _1_.b);
}
"
`;

exports[`Optional chaining Corner cases should await expression 2 1`] = `
"var _1_;
async function main() {
  await((_1_ = a) === null || _1_ === void 0 ? void 0 : _1_.b());
}
"
`;

exports[`Optional chaining Corner cases should await expression with ??  1`] = `
"var _1_;
var _2_;
let b = async (a, c) => {
  const b = await((_1_ = a) === null || _1_ === void 0 ? void 0 : _1_.a()) ?? await((_2_ = c) === null || _2_ === void 0 ? void 0 : _2_.d);
};
"
`;

exports[`Optional chaining Corner cases should handle asExpression 1`] = `
Object {
  "code": "var _1_, _2_;
var _3_, _4_;
(_2_ = (_1_ = a) === null || _1_ === void 0 ? void 0 : _1_.b) === null || _2_ === void 0 ? void 0 : _2_.c;
(_4_ = (_3_ = a) === null || _3_ === void 0 ? void 0 : _3_.b) === null || _4_ === void 0 ? void 0 : _4_.c;
",
  "requireStatementCollection": Array [],
}
`;
